src/ast/nodes/shared/FunctionNode.ts

```diff
@@ -16,12 +16,16 @@ export default class FunctionNode extends NodeBase {
 	id!: IdentifierWithVariable | null;
 	params!: PatternNode[];
 	preventChildBlockScope!: true;
+	referencesThis!: boolean;
 	scope!: FunctionScope;
 
 	private isPrototypeDeoptimized = false;
 
 	createScope(parentScope: FunctionScope) {
 		this.scope = new FunctionScope(parentScope, this.context);
+		// Initialized here because child nodes will update it before
+		// `this.initialize` even runs.
+		this.referencesThis = false;
```

Comment:
Ah I see. Maybe move it rather into `parseNode`, which is where "initialisation before initialise" usually happens. I.e. `parseNode` is guaranteed to run before child nodes are even created.